pkg/openfeature/api.go

```diff
@@ -0,0 +1,170 @@
+package openfeature
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/open-feature/go-sdk/pkg/openfeature/internal"
+)
+
+// evaluationAPI wraps OpenFeature evaluation API functionalities
+type evaluationAPI struct {
+	defaultProvider FeatureProvider
+	namedProviders  map[string]FeatureProvider
+	hks             []Hook
+	evalCtx         EvaluationContext
+	logger          logr.Logger
+	mu              sync.RWMutex
+}
+
+// newEvaluationAPI is a helper to generate an API. Used internally
+func newEvaluationAPI() evaluationAPI {
+	return evaluationAPI{
+		defaultProvider: NoopProvider{},
+		namedProviders:  map[string]FeatureProvider{},
+		hks:             []Hook{},
+		evalCtx:         EvaluationContext{},
+		logger:          logr.New(internal.Logger{}),
+		mu:              sync.RWMutex{},
+	}
+}
+
+// setProvider sets the default FeatureProvider of the evaluationAPI. Returns an error if FeatureProvider is nil
+func (api *evaluationAPI) setProvider(provider FeatureProvider) error {
+	api.mu.Lock()
+	defer api.mu.Unlock()
+
+	if provider == nil {
+		return errors.New("default provider cannot be set to nil")
+	}
+
+	// Initialize new default provider and shutdown the old one
+	// Provider update must be non-blocking, hence initialization & shutdown happens concurrently
+	oldProvider := api.defaultProvider
+	go func() {
```

Comment:
maybe worth having a dedicate `func switchProvider(old, new FeatureProvider)` function that is shared with setNamedProvider

Reply:
Good suggestion. I will add this with the next commit

Reply:
Logic is now isolated and shared with callers